src/Stratis.Bitcoin.Features.PoA/PoAMiner.cs

```diff
@@ -179,7 +168,41 @@ private async Task CreateBlocksAsync()
             }
         }
 
-        protected virtual async Task WaitBeforeCanMineAsync(int delayMs, CancellationToken cancellation = default(CancellationToken))
+        /// <summary>
+        /// Doesn't let the node progress until the given time slot.
+        /// </summary>
+        /// <param name="myTimestamp">The next time slot this node should mine at.</param>
+        protected virtual async Task WaitUntilWeCanMineAsync(uint myTimestamp)
+        {
+            int waitingTime = this.GetWaitingTimeInSeconds(myTimestamp);
+
+            // Accounts for different IDateTimeProvider implementations. The standard implementation would progress
+            // after one iteration, but for other cases where we are emulating time, this may loop more than once.
+            while (waitingTime > 0)
+            {
+                this.logger.LogInformation("Waiting {0} seconds until block can be mined.", waitingTime);
+
+                await this.TaskDelayAsync(waitingTime * 1000, this.cancellation.Token).ConfigureAwait(false);
+
+                waitingTime = this.GetWaitingTimeInSeconds(myTimestamp);
```

Comment:
why do we need to recalculate this here?

Reply:
`GetWaitingTimeInSeconds` is what hits the `IDateTimeProvider`, so we get the new time and either continue out if it's 0, or keep waiting until the `IDateTimeProvider` returns a different result.